### PR TITLE
buildSupport: add helpers for verifying code signatures

### PR DESCRIPTION
**Pull request by xeji**
###### Motivation for this change

Many projects provide signed source tarballs or signed binaries, which are currently not checked in nixpkgs. While it's not perfect, I believe that checking signatures provides some security advantage our typical "TOFU" workflow of implicitly trusting changed sha256 hashes.

The main reason is that, as opposed to sha256 hashes, public keys are usually long term, so we can invest more time to verify them initially, and after doing this once, malicious changes are more likely to be spotted.

This PR proposes some simple helpers for checking code signatures. It also provides some examples on how to use them.

###### Things done

- `fetchpgpkey` is an extension of `fetchurl` that fetches a PGP public key to the nix store and verifies its fingerprint with `gpg`. It still needs a sha256 because it is based on `fetchurl`.  (I have experimented with an alternative implementation using `builtins.fetchurl` and just a fingerprint, no sha256, but that is not strictly a fixed-output derivation).

- `verifySignatureHook` sets up signature verification and adds a `preUnpackHook` to automatically verify the signature `srcSignature` of `src` with `signaturePublicKey` when present. Verification can also be triggered manually using the shell functions defined in the hook.

- Implemented 3 different examples for various use cases of the hook. These examples are for demonstartion and may or may not be merged with this PR. 

    - `samba4`: the tarball is signed, but the signature is for the uncompressed tar archive. The hook has an option for this.
    - `tor-browser-bundle-bin`: provides signed tarballs but there's not standard unpack phase, so the verification must be triggered manually in `buildCommand`
    -  `1password`: the tarball contains a signed binary and the signature file, so verification is done in `checkPhase` here.

###### Possible future work

Enhance `fetchgit` to verify signed revisions using `git verify-commit`.

---

cc @matthewbauer @jb55 @garbas

### CODE REVIEW

File: pkgs/servers/samba/4.x.nix [hunk 1]
```diff
@@ -29,6 +31,20 @@ stdenv.mkDerivation rec {
...
+  };
+  signatureUncompressed = true;
+
+  signaturePublicKey = fetchpgpkey {
+    url = https://download.samba.org/pub/samba/samba-pubkey.asc;
+    sha256 = "1fndhq0c34va34z137gvsl9gpwjv30b06makfx8cq5vrmgiax1x1";
+    fingerprint = "52FBC0B86D954B0843324CDC6F33915B6568B7EA";
+  };
+
+  nativeBuildInputs = [ verifySignatureHook ];
```
Comment (matthewbauer) [inline pkgs/servers/samba/4.x.nix]: Does this work? We aren't downloading binaries here so I have no idea where the signature would be coming from. This should be not needed at all for source based builds.

**Comment by orivej** [comment]
It should be worth taking a look at how Triton did it: see https://github.com/triton/triton/commits/master?before=6a1de201948b7755177562105e49181baca2bc2e+35&path[]=pkgs&path[]=build-support&path[]=fetchurl&path[]=builder.sh starting from Mar 30, 2016. (/cc @wkennington in case you are interested or would like to help.)

File: pkgs/servers/samba/4.x.nix (hunk 1 above)
Comment (xeji) [reply pkgs/servers/samba/4.x.nix]: It works by convention: The hook assumes that  `srcSignature` contains the signature file, and `signaturePublicKey` the corresponding key. If they're present, `srcSignature` is checked against `src`.  Of course you have to `fetchurl` the signature file and `fetchpgpkey` the key for the hook to work.

Comment (xeji) [reply pkgs/servers/samba/4.x.nix]: And yes, it can be used for source base builds if upstream provides a signature for the source tarball. It obviously doesn't work with `fetchFromGitHub`, but that would be a case for verifying signed git commits.

**Comment by matthewbauer** [comment]
Okay a few thoughts on this.

1. I think there is a fairly good case to be made to do this for 1password. We are relying on a binary blob & that's unfortunate. I know a lot of people do use 1password & there is definitely good reason to support it in Nixpkgs because of that. However, we should be encouraging everyone to use free and open source software as much as possible! Nixpkgs just does not work very well with binary blobs.
2. The setup hook makes a lot more sense to me - it is much more reusable & we now don't have to check in the public key. I don't really understand the need for two layers of integrity checks but enough people have voiced support that maybe it is worth doing. At least I don't think it will hurt to do this.
3. I would strongly encourage us to drop the tor-browser-bundle-bin & firefox-bin. It makes no sense to support these things when we have a good source-based version of each of them. If someone wants to maintain them please put them in something like NUR or another overlay. Nix has never worked well with binary blobs and this is partly by design. This isn't exactly related to this PR but just something that I had forgotten was still around.

File: pkgs/servers/samba/4.x.nix (hunk 1 above)
Comment (matthewbauer) [reply pkgs/servers/samba/4.x.nix]: Ok did not realize that on tarball signatures! I wonder if we could put the source-based part of the hook in fetchurl then? Just have an optional attribute that is verified for us. This would also make it possible to support multiple-src derivations.

**Comment by xeji** [comment]
> However, we should be encouraging everyone to use free and open source software as much as possible! Nixpkgs just does not work very well with binary blobs.

@matthewbauer fully agree.  I just randomly selected these examples to test and show different use cases of the setup hook.

**Comment by xeji** [comment]
Marked WIP while I have a look at how Triton does it and try to put part of the features into `fetchurl`.

**Comment by volth** [comment]
> It still needs a ```sha256``` because it is based on ```fetchurl```. (I have experimented with an alternative implementation using ```builtins.fetchurl``` and just a fingerprint, no ```sha256```, but that is not strictly a fixed-output derivation).

If you move the verification from ```buildPhase``` to ```postFetch``` of a fixed-output derivation performing the download (one of ```fetchurl```/```fetchzip```/..), ```fetchgpgkey``` could be an [impure derivation](https://github.com/NixOS/nix/commit/647291cd6c7559f68d49a5cdd907c2fd580790b1) without ```sha256```.

**Comment by jb55** [comment]
Would it be hard to support this common use case? Here's are the instructions for verifying bitcoin signatures:

> Verify that the checksum of the release file is listed in the checksums
> file using the following command:

    sha256sum --ignore-missing --check SHA256SUMS.asc

> In the output produced by the above command, you can safely ignore any
> warnings and failures, but you must ensure the output lists "OK" after
> the name of the release file you downloaded. For example:

    bitcoin-0.16.2-x86_64-linux-gnu.tar.gz: OK

> Obtain a copy of the release signing key by running the following command:

    gpg --recv-keys 01EA5486DE18A882D4C2684590C8019E36C2E964

> The output of the command above should say that one key was imported,
> updated, has new signatures, or remained unchanged.
> Verify that the checksums file is PGP signed by the release signing key:

    gpg --verify SHA256SUMS.asc

**Comment by c0bw3b** [comment]
Ooooh I would really like to see this happen.
I don't see it as an extra layer of integrity checking. The SHA checksum ensure integrity of the file but verifying the file signature ensure the identity of the source. There is _some_ overlap in practical but it remains two different goals.

I'm discovering this PR just now and it seems the proposed implementation is easily extensible.
A future improvement could be to support verification of Ed25519 signatures (think `minisign` or OpenBSD `signify`). It seems it would be easy enough to write `fetched25519key` and then extend `verifySignatureHook` to check whether it's fed with a PGP or Ed25519 key.

TL;DR; I'll now read other posts

**Comment by jb55** [comment]
> A future improvement could be to support verification of Ed25519 signatures 

gpg supports ed25519 sigs

**Comment by joachifm** [comment]
Not to get into whether tor-browser-bundle-bin should continue to live, but I want to note that its signature is already verified via the updater.

File: pkgs/build-support/setup-hooks/verify-signature.sh [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+
+# verifySignature SIGFILE DATAFILE UNCOMPRESS
+# verify the signature SIGFILE for the file DATAFILE
+# if DATAFILE is omitted, it is derived from SIGFILE by dropping the .asc or .sig suffix
+# if UNCOMPRESS is set, uncompress DATAFILE before verification
+verifySignature() {
+  if [ -z "$3" ]; then
+    gpgv --keyring pubring.kbx "$1" "$2" || exit 1
```
Comment (Mic92) [inline pkgs/build-support/setup-hooks/verify-signature.sh]: Can you make gpg ignoring expire dates? Otherwise it will become a problem when somebody tries to build an old nixpkgs revision. We could print a warning if a gpg key is expired.

Comment (Mic92) [reply pkgs/build-support/setup-hooks/verify-signature.sh]: The warning should be good enough for the TOFU case.

File: pkgs/applications/misc/1password/default.nix [hunk 3]
```diff
@@ -24,6 +24,19 @@ stdenv.mkDerivation rec {
       }
     else throw "Architecture not supported";
 
+  nativeBuildInputs = [ verifySignatureHook ];
+
+  signaturePublicKey = fetchpgpkey {
+    url = https://keybase.io/1password/pgp_keys.asc;
```
Comment (Mic92) [inline pkgs/applications/misc/1password/default.nix]: What if they update their key?

Comment (Mic92) [reply pkgs/applications/misc/1password/default.nix]: Can we fetch their key from a key-server using the current keyid instead?

Comment (Ekleog) [reply pkgs/applications/misc/1password/default.nix]: Keyservers are just one of many ways of fetching OpenPGP keys, I don't think it would make sense to force downloads through them (also, we'd want to fetch using the fingerprint, not the keyid).

Having a `fetchFromSks` wrapper of `fetchpgpkey` that fetches from the sks-keyservers network might be a meaningful addition, but I don't think it's required for the first iteration

File: pkgs/build-support/fetchpgpkey/default.nix [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+{
+  fingerprint
+, ... } @ args:
+
+lib.overrideDerivation (fetchurl ({
+
+  name = "pubkey-${fingerprint}";
+
+  postFetch =
```
Comment (Ekleog) [inline pkgs/build-support/fetchpgpkey/default.nix]: I think we could reduce issues when a key gets an additional signature (eg. for when fetching from sks-keyservers) by stripping it of all UIDs, signatures, etc. before the `sha256` check.

**Comment by roconnor-blockstream** [comment]
Please do not make this part of `fetchurl`.  Placing it as part of `fetchurl` would enable an attack where and attacker tricks a victim to load a trojan-horse application into their nix-store and then updates the hash of a fetchurl in `nixpkgs` to reference the trojan horse.

Now I do respect that if the attacker has access to the victim's `nixpkgs`, there are a lot of bad things they can do, but getting limited "access" to `nixpkgs` hashes is maybe easier that you might think (see the reference to `r-ryantm` below).  However my larger point is that putting a GPG check into the derivation "guarantees" that it has been executed during the build, but putting the check in `fetchurl` does not, because `fetchurl` is simply bypassed if a file with the listed hash value ever already exists in the store.

As to motivation of why we want this: We'd like to use automatic updates by bots such as `r-ryantm`, however an attacker can currently attack packages by hijacking domains and serving "updated" trojan-horse applications, and get nixpkgs to be (semi-)automatically updated.  For those packages with signed source releases, it would let us use automatic updates for signed software more safely.

**Comment by Mic92** [comment]
I agree with @roconnor-blockstream we should make this part of our tooling rather than the build process. Once we have our own checksum we no longer need to rely on public key cryptography.

**Comment by danielfullmer** [comment]
@Mic92  Correct me if otherwise, but I don't think @roconnor-blockstream was saying that it shouldn't be verified via Nix in a derivation, only that the verification shouldn't be behind a fixed output derivation like `fetchurl`.  Having a signature check take place in Nix provides assurance of build input _authenticity_ in addition to just _integrity_.

There is also another alternative I haven't seen mentioned:  Have another derivation that takes the source and signature as input, and outputs a symlink to the original source if and only if it passes a signature check.  The output of that verification derivation could then be passed as the `src` to `mkDerivation`.  This seems like more clean separation of concerns, rather than just adding signature checking as an additional setup hook in the stdenv. But, it comes at the cost of an additional derivation.  I have an example use of this idea for android APKs here: https://github.com/danielfullmer/robotnix/pull/59
